server/base/src/main/java/org/apache/accumulo/server/metrics/PerTableMetrics.java

```diff
@@ -0,0 +1,203 @@
+   *        passed for consistency with
+   *        {@link #newPerTableMetrics(MeterRegistry, TableId, Consumer, List)}
+   * @param tags currently an empty collection of tags, this is passed for consistency with
+   *        {@link #PerTableMetrics(ServerContext, Supplier)}
+   * @return a new object that will be cached and later returned by
+   *         {@link #getTableMetrics(TableId)}
+   */
+  protected abstract T newAllTablesMetrics(MeterRegistry registry, Consumer<Meter> meters,
+      List<Tag> tags);
+
+  /**
+   *
+   * Subclasses should implement this method to create per table table metrics objects. This method
+   * is called in the case where per table metrics are enabled. These objects will be cached and
+   * returned by {@link #getTableMetrics(TableId)}. Table metrics object in the cache that are no
+   * longer needed will be automatically removed when the table is deleted or this server has not
+   * hosted the table for a bit.
+   *
+   * @param registry register an meters for the table metrics in this registry
+   * @param meters a consumer that accepts meters to be removed from the registry when the per table
+   *        metrics object is discarded.
+   * @param tags returns a list with a single tag in it which is the tableId. These tags should be
+   *        used when registering meters
+   * @return a new object that will be cached and later returned by
+   *         {@link #getTableMetrics(TableId)}
+   */
+  protected abstract T newPerTableMetrics(MeterRegistry registry, TableId tableId,
+      Consumer<Meter> meters, List<Tag> tags);
+
+  private TableMetricsInfo<T> getOrCreateTableMetrics(TableId tableId) {
+    Preconditions.checkState(perTableActive);
+    return perTableMetrics.computeIfAbsent(tableId, tid -> {
+      List<Meter> meters = new ArrayList<>();
+      T tableMetrics = newPerTableMetrics(registry, tableId, meters::add,
+          List.of(Tag.of(TABLE_ID_TAG_NAME, tid.canonical())));
+      getLog().debug("Created {} meters for table id {} in metrics registry.", meters.size(),
+          tableId);
+      return new TableMetricsInfo<>(tableMetrics, meters);
+    });
+  }
+
+  public void registerMetrics(MeterRegistry registry) {
+    Preconditions.checkState(this.registry == null);
+    this.registry = registry;
+    if (!perTableActive) {
+      this.allTableMetrics = newAllTablesMetrics(registry, m -> {}, List.of());
+    }
+  }
+
+  public T getTableMetrics(TableId tableId) {
+    Preconditions.checkState(registry != null);
+
+    if (!perTableActive) {
+      return allTableMetrics;
+    }
+
+    return getOrCreateTableMetrics(tableId).tableMetrics;
+  }
+
+  /**
+   * This method will create per table metrics for any tables that are active on this server and
+   * currently have no table metrics object in the cache. It will also remove an per table metrics
+   * object from the cache that have been inactive for a while or where the table was deleted.
+   */
+  public synchronized void refresh() {
```

Comment:
I'm curious if it might be better to evaluate whether per table metrics should be added or removed when a tablet is hosted or unhosted in the ScanServer and TabletServer. There are explicit mechanisms in the TabletServer for hosting and unhosting tablets. In the ScanServer we have the TabletMetadataLoader for hosting a tablet, and we could add an [evictionListener](https://github.com/ben-manes/caffeine/wiki/Removal) to the tabletMetadataCache to handle a tablet removal. Thoughts on that?

Reply:
It probably would be better to do that for all cases.  Its only being done for the tablet server on tablet load for this  method.  The three other cases you mentioned are not done.  

For the scan server I could not find a good place to register on tablet load, I will circle back and see what I can find.  For now its probably ok that scan server does not register on load because it has no gauges, so when a scan happen it will touch meters which will load metrics.  However that is shaky ground, if gauges were ever used then those may not be loaded until the timer task kicks in.  Would also be good to push code to TabletHostingServer so that the metrics code can interact w/ the same code for each server type.


If all 4 cases are covered with callbacks then we could run the timer task less frequently.  For the unload case I was completely leaving that to the timer task to catch.

Reply:
Made some changes related to this in afa401f.  Was able to optimize and centralize the code for detecting changes in the set of table ids. Using those changes could efficiently handle a tablet being loaded and detect if anything needed to be done.   However for the case of a tablet being unloaded found that is hard to handle that efficiently because when one tablet is unloaded other tablet may still have that same tablet id, so need to scan all tablets on each tablet unload to see if anything needs to be done.  Decided not to do anything for this case and leave it to the periodic timer task.  Was able to centralize that timer task and make it more efficient though.